### PR TITLE
Cleanup cimports, implement bits of numpy_helper in util.pxd

### PR DESCRIPTION
**Pull request by jbrockmendel**
Removes a few unnecessary uses of `cnp.import_array()`

Uses cython's conditional compilation to avoid runtime PY2/PY3 checks

Arranges cython imports in stdlib-->3rd party--> internal order.

Separates out the one part of `libresolution._FrequencyInferer` that we actually _do_ want to keep in cython

Couple of small lintings in groupby.pyx

### CODE REVIEW

File: pandas/_libs/index.pyx [hunk 1]
```diff
@@ -9,14 +9,12 @@ from cpython.slice cimport PySlice_Check
 import numpy as np
 cimport numpy as cnp
 from numpy cimport (ndarray, float64_t, int32_t,
-                    int64_t, uint8_t, uint64_t, intp_t)
+                    int64_t, uint8_t, uint64_t, intp_t,
+                    # Note: NPY_DATETIME, NPY_TIMEDELTA are only available
```
Comment (jreback) [inline pandas/_libs/index.pyx]: prob can remove this comment (future PR ok)

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -54,7 +52,8 @@ from tslibs.timestamps cimport (create_timestamp_from_ts,
                                 _NS_UPPER_BOUND, _NS_LOWER_BOUND)
 from tslibs.timestamps import Timestamp
 
-cdef bint PY2 = str == bytes
+
+DEF PY2 = str == bytes
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: we should define in util.pxd maybe (future ok)

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -334,8 +334,6 @@ class _BaseOffset(object):
             # other is not a DateOffset object
             return False
 
-        return self._params == other._params
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: this didn't break anything?

File: pandas/_libs/tslibs/period.pyx [hunk 4]
```diff
@@ -46,14 +46,14 @@ from conversion cimport tz_convert_utc_to_tzlocal
...
-from parsing import parse_time_string, NAT_SENTINEL
+from parsing import parse_time_string
 from resolution import Resolution
 from nattype import nat_strings, NaT, iNaT
 from nattype cimport _nat_scalar_rules, NPY_NAT, is_null_datetimelike
 from offsets cimport to_offset
 from offsets import _Tick
 
-cdef bint PY2 = str == bytes
+DEF PY2 = str == bytes
```
Comment (jreback) [inline pandas/_libs/tslibs/period.pyx]: same

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel a couple of things ought to do in the future PRs

File: pandas/_libs/tslibs/offsets.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: Was unreachable; there was a compile-time warning about it.

File: pandas/_libs/tslib.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx]: I doubt that works for DEF values, but not much downside.

**Pull request merged**